l2-contracts/contracts/bridge/L2SharedBridge.sol

```diff
@@ -39,169 +40,172 @@ contract L2SharedBridge is IL2SharedBridge, Initializable {
     address public override l1Bridge;
 
+    /// @dev Chain ID of Era for legacy reasons
     uint256 public immutable ERA_CHAIN_ID;
 
+    /// @dev Chain ID of L1 for bridging reasons
+    uint256 public immutable L1_CHAIN_ID;
+
+    IL2NativeTokenVault public nativeTokenVault;
+
+    /// @dev A mapping l2 token address => l1 token address
+    mapping(bytes32 assetId => address assetHandlerAddress) public override assetHandlerAddress;
+
+    /// @notice Checks that the message sender is the legacy bridge.
+    modifier onlyL1Bridge() {
+        // Only the L1 bridge counterpart can initiate and finalize the deposit.
+        if (
+            AddressAliasHelper.undoL1ToL2Alias(msg.sender) != l1Bridge &&
+            AddressAliasHelper.undoL1ToL2Alias(msg.sender) != l1SharedBridge
+        ) {
+            revert InvalidCaller(msg.sender);
+        }
+        _;
+    }
+
     /// @dev Contract is expected to be used as proxy implementation.
     /// @dev Disable the initialization to prevent Parity hack.
-    constructor(uint256 _eraChainId) {
+    constructor(uint256 _eraChainId, uint256 _l1ChainId) {
         ERA_CHAIN_ID = _eraChainId;
+        L1_CHAIN_ID = _l1ChainId;
         _disableInitializers();
     }
 
     /// @notice Initializes the bridge contract for later use. Expected to be used in the proxy.
     /// @param _l1SharedBridge The address of the L1 Bridge contract.
     /// @param _l1Bridge The address of the legacy L1 Bridge contract.
-    /// @param _l2TokenProxyBytecodeHash The bytecode hash of the proxy for tokens deployed by the bridge.
-    /// @param _aliasedOwner The address of the governor contract.
+    /// @param _assetHandler The address of the nativeTokenVault contract.
     function initialize(
         address _l1SharedBridge,
         address _l1Bridge,
-        bytes32 _l2TokenProxyBytecodeHash,
-        address _aliasedOwner
+        IL2NativeTokenVault _assetHandler
     ) external reinitializer(2) {
         if (_l1SharedBridge == address(0)) {
             revert EmptyAddress();
         }
-
-        if (_l2TokenProxyBytecodeHash == bytes32(0)) {
-            revert EmptyBytes32();
-        }
-
-        if (_aliasedOwner == address(0)) {
+        if (address(_assetHandler) == address(0)) {
             revert EmptyAddress();
         }
 
         l1SharedBridge = _l1SharedBridge;
-
-        if (block.chainid != ERA_CHAIN_ID) {
-            address l2StandardToken = address(new L2StandardERC20{salt: bytes32(0)}());
-            l2TokenBeacon = new UpgradeableBeacon{salt: bytes32(0)}(l2StandardToken);
-            l2TokenProxyBytecodeHash = _l2TokenProxyBytecodeHash;
-            l2TokenBeacon.transferOwnership(_aliasedOwner);
-        } else {
+        nativeTokenVault = _assetHandler;
+        if (block.chainid == ERA_CHAIN_ID) {
             if (_l1Bridge == address(0)) {
                 revert EmptyAddress();
             }
             l1Bridge = _l1Bridge;
```

Comment:
not sure we need to always rewrite this value for Era. I would not do it unless the `l1Bridge` is already zero. Otherwise we create risk of breaking era with each such upgrade

Reply:
good point